### PR TITLE
Fix out of memory issue by disabling syntax tree caching 

### LINKED ISSUE #40398: [Improvement]: Improve EnvironmentPackageCache to Cache Only Required Data
### Description

Currently, all the data, including syntax trees and documents, is cached in the project API for later use. However, when dealing with a large dependency tree, this can result in out-of-memory errors (OOM). Initially, these data are cached to be used by LS and other tools. However, it has been observed that LS does not utilize this cache and instead relies on the Compiler API to derive the required information. Even the resolution of 'Go to Definition' is handled differently. After discussions with @azinneera and @IMS94, we concluded that it would be worth exploring the possibility of cleaning up unnecessary data once the BIR  is generated.

Here is a heap dump generated for the following Ballerina program. 

```
import ballerina/io;
import wso2healthcare/healthcare.fhir.r4;
import wso2healthcare/healthcare.hl7v23;
import wso2healthcare/healthcare.hl7v231;
import wso2healthcare/healthcare.hl7v24;
import wso2healthcare/healthcare.hl7v25;
import wso2healthcare/healthcare.hl7v251;
import wso2healthcare/healthcare.hl7v26;
```

Screenshot 2023-05-10 at 16 34 48

### Describe your problem(s)

_No response_

### Describe your solution(s)

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by Thevakumar-Luheerathan**
Fixes #40398
Fixes https://github.com/wso2-enterprise/internal-support-ballerina/issues/321

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/RunCommand.java [hunk 1]
```diff
@@ -254,7 +258,8 @@ private BuildOptions constructBuildOptions() {
                 .setSticky(sticky)
                 .setDumpGraph(dumpGraph)
                 .setDumpRawGraphs(dumpRawGraphs)
-                .setConfigSchemaGen(configSchemaGen);
+                .setConfigSchemaGen(configSchemaGen)
+                .disableSyntaxTree(disableSyntaxTreeCaching);;
```
Comment (gayaldassanayake) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/RunCommand.java:262]:
```suggestion
                .disableSyntaxTree(disableSyntaxTreeCaching);
```

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/DocCommand.java [hunk 2]
```diff
@@ -214,7 +218,8 @@ private BuildOptions constructBuildOptions() {
                 .setCodeCoverage(false)
                 .setOffline(offline)
                 .setTestReport(false)
-                .setObservabilityIncluded(false);
+                .setObservabilityIncluded(false)
+                .disableSyntaxTree(disableSyntaxTreeCaching);;
```
Comment (gayaldassanayake) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/DocCommand.java:222]:
```suggestion
                .disableSyntaxTree(disableSyntaxTreeCaching);
```

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/PackageConfigCreator.java [hunk 3]
```diff
@@ -52,7 +52,7 @@
  */
 public class PackageConfigCreator {
 
-    public static PackageConfig createBuildProjectConfig(Path projectDirPath) {
+    public static PackageConfig createBuildProjectConfig(Path projectDirPath, Boolean disableSyntaxTree) {
```
Comment (gayaldassanayake) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/PackageConfigCreator.java:55]: a `boolean` data type will be sufficient here.

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/PackageConfigCreator.java [hunk 4]
```diff
@@ -89,9 +95,15 @@ public static PackageConfig createSingleFileProjectConfig(Path filePath) {
...
-        return createPackageConfig(packageData, packageManifest, dependencyManifest);
+        return createPackageConfig(packageData, packageManifest, dependencyManifest, DependencyGraph.emptyGraph(),
+                Collections.emptyMap(), !disableSyntaxTree);
+    }
+
+    public static PackageConfig createSingleFileProjectConfig(Path filePath) {
+        return createSingleFileProjectConfig(filePath, false);
     }
 
+
```
Comment (gayaldassanayake) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/PackageConfigCreator.java:106]:
```suggestion
```

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/BuildOptions.java [hunk 5]
```diff
@@ -71,6 +73,10 @@ public boolean sticky() {
         return this.compilationOptions.sticky();
     }
 
+    public boolean disableSyntaxTree() {
```
Comment (azinneera) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/BuildOptions.java:76]: Let's make this a compilation option. BuildOptions will be suggested in the Ballerina.toml in VSCode AFAIR.

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/DocumentContext.java [hunk 6]
```diff
@@ -59,15 +59,29 @@ class DocumentContext {
...
     private final String content;
+    private boolean enableSyntaxTree = true;
 
     private DocumentContext(DocumentId documentId, String name, String content) {
         this.documentId = documentId;
         this.name = name;
         this.content = content;
     }
 
+    private DocumentContext(DocumentId documentId, String name, String content, boolean enableSyntaxTree) {
```
Comment (azinneera) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/DocumentContext.java:70]: Let's modify the existing constructor without introducing a new one.

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/ModuleCompilationState.java [hunk 7]
```diff
@@ -28,7 +28,6 @@ enum ModuleCompilationState {
     LOADED_FROM_SOURCES {
         @Override
         void parse(ModuleContext moduleContext) {
-            ModuleContext.parseInternal(moduleContext);
```
Comment (azinneera) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/ModuleCompilationState.java:31]: Why have we removed this?

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/BuildCommand.java [hunk 8]
```diff
@@ -299,7 +303,10 @@ private BuildOptions constructBuildOptions() {
                 .setExportOpenAPI(exportOpenAPI)
                 .setExportComponentModel(exportComponentModel)
                 .setEnableCache(enableCache)
-                .setNativeImage(nativeImage);
+                .setNativeImage(nativeImage)
+                .disableSyntaxTree(disableSyntaxTreeCaching);
+
+
```
Comment (azinneera) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/BuildCommand.java:309]: unnecessary newline

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/DocumentContext.java [hunk 9]
```diff
@@ -59,15 +59,29 @@ class DocumentContext {
     private final DocumentId documentId;
     private final String name;
     private final String content;
+    private boolean enableSyntaxTree = true;
```
Comment (azinneera) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/DocumentContext.java:62]: The name is conflicting with the input. Let's rename this to `disableSyntaxTree`

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/TestCommand.java [hunk 10]
```diff
@@ -360,7 +364,8 @@ private BuildOptions constructBuildOptions() {
                 .setDumpGraph(dumpGraph)
                 .setDumpRawGraphs(dumpRawGraphs)
                 .setNativeImage(nativeImage)
-                .setEnableCache(enableCache);
+                .setEnableCache(enableCache)
+                .disableSyntaxTree(disableSyntaxTreeCaching);
```
Comment (azinneera) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/TestCommand.java:368]:
```suggestion
                .disableSyntaxTreeCaching(disableSyntaxTreeCaching);
```
Let's rename it.

**Review by azinneera — APPROVED** [review]

**Pull request merged**